### PR TITLE
Fix the navigation issue inside cover blocks

### PR DESCRIPTION
**Pull request by renatho**
## What?

Previously, we implemented [this fix](https://github.com/WordPress/gutenberg/pull/65626), but [a regression was found](https://github.com/WordPress/gutenberg/issues/66067), so it was [reverted](https://github.com/WordPress/gutenberg/pull/66074), and we are sending this new fix using a different approach.

## Why?

Many users are having and reporting this issue.
See
- https://github.com/WordPress/gutenberg/issues/45353
- https://github.com/Automattic/wp-calypso/issues/75439
- https://github.com/Automattic/wp-calypso/issues/79317

The reason is that the cover inner blocks container contains a z-index, crating a stacking context for what is inside.

## How?

The new solution changes the `z-index` of the cover inner blocks container to `auto` when the navigation is open (`has-modal-open` applied to the `html`). I don't consider it as the ideal solution, but it's the simplest and probably the safest one.

What I would see as "ideal" (best in terms of good practices of code if we were creating it from scratch) solutions (and the reason that I didn't implement them):

1. Having the menu opened in a totally different context would be the ideal solution to avoid any stacking context issues. But besides it being a big change, it could impact other things, like not inheriting styles from the current context and creating other unexpected style issues.
2. Change the order of the cover elements (just invert the image and the color effect span). With this, we could remove the `z-index` of the elements, and it would solve the issue. The problem would be the block deprecation and a complex code to maintain both versions to not break the styles on the frontend until the user re-saves the cover block.

## Testing Instructions

1. In a template or a post add a Cover block.
2. Inside the cover block, add a Navigation block.
3. After the previously created Cover block, add another Cover block.
4. Visit the site in the frontend, on a small screen.
5. Open the menu, and make sure it's displayed over the whole content.

cc @aaronrobertshaw in case you could re-test it.

## Screenshots or screencast 

https://github.com/user-attachments/assets/98bdfda3-4608-4869-90b1-887b09eaab89

### CODE REVIEW

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (renatho) [inline test/e2e/specs/editor/blocks/cover.spec.js]: > On the topic of tests, given the complexities around when the cover block should and shouldn't be visible due to z-index styling, it would be great if we could have some tests to protect against future regressions.

@aaronrobertshaw, this was a bit challenging. Basically, I couldn't find a way to check if an element would be over the other. It could be a screenshot, but it seems it's an approach we are not using in Gutenberg tests and it could easily change.

But I created a helper function (`calculateZIndexContext`) to get the zIndex of the context where the element is. It's also not perfect because if we have an element parent of both blocks with a `z-index`, both would have the same because we're not comparing in what level of the tree they are. Since we aren't testing it with other blocks as wrapper, it would be fine unless at some point we decide to add some wrappers to the editor with `z-index`, but I suspect we should be fine here.

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/cover.spec.js]: Would `toBeVisible` be sufficient here? I'm not sure what the main goal of this test is. I wonder if we could test this from the accessibility perspective. If a modal is opened, then all the other elements behind it should be inaccessible and [hidden](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-hidden). I don't think checking the `z-index` is good either since it's an implementation detail; having a higher `z-index` doesn't mean it's top in the stacking context.

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/cover.spec.js]: Is there an accessible target we can test against instead of using implementation details like CSS class selectors? `getByRole` is generally preferred when selecting elements.

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (aaronrobertshaw) [reply test/e2e/specs/editor/blocks/cover.spec.js]: Thanks for weighing in @kevin940726 🙇 

> Would toBeVisible be sufficient here? 

My understanding of the [Playwright docs](https://playwright.dev/docs/actionability#visible) are that `toBeVisible` or `toBeHidden` won't quiet be enough as they don't take z-index into account at all.

> Element is considered visible when it has non-empty bounding box and does not have visibility:hidden computed style.
> 
> Note that according to this definition:
> 
> Elements of zero size are not considered visible.
> Elements with display:none are not considered visible.
> Elements with opacity:0 are considered visible.

When I initially suggested a test here I was hoping we could get away with something like `toBeVisible` to avoid testing the implementation details. Given that isn't straightforward, I don't think we want this test to block a fix landing. 

Sorry to have sent you on a wild goose chase @renatho 🙏 

> If a modal is opened, then all the other elements behind it should be inaccessible and [hidden](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-hidden).

There's also another issue where certain elements/blocks behind the navigation overlay can still be interacted with.  Could that throw another spanner in the works?

That said, it does spark another possible idea. Maybe we could do a [`.click( { trial: true } )`](https://playwright.dev/docs/api/class-locator#locator-click-option-trial) on the elements in question. It should return true if the element can be clicked and false if it's overlapped?  As the nav overlay should fill the page and completely cover the hidden block, I'm thinking we wouldn't need to worry about a position for the click.

Take that with a grain of salt. It could be an odd suggestion 🤷 

The Playwright docs include this comment within its [actionability check details](https://playwright.dev/docs/actionability):

> element [Receives Events](https://playwright.dev/docs/actionability#receives-events), as in not obscured by other elements

File: packages/block-library/src/navigation/edit/index.js
Comment (aaronrobertshaw) [inline packages/block-library/src/navigation/edit/index.js]: Nit: I'm not sure this inline comment adds much value. Is it worth expanding on or removing?

File: packages/block-library/src/navigation/edit/index.js
Comment (aaronrobertshaw) [inline packages/block-library/src/navigation/edit/index.js:88]: It does feel a bit odd for a block to be manipulating the root node but if this brings it inline with the frontend perhaps it's already been decided this is ok?

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for your patience here @renatho 👍 

This PR branch needs rebasing on trunk to pick up the revert of the original approach otherwise the cover styles are broken.

After rebasing locally, I've given this another test and I think it works well.

| Trunk | This PR |
|---|---|
| <video src="https://github.com/user-attachments/assets/e64a3bb9-261a-487b-987b-4e0d98a4b036" /> | <video src="https://github.com/user-attachments/assets/73adae88-1ed0-46d1-9eac-252c08054af8" /> |

Given the false start already with the original PR, I'd like to get a second set of eyes on this latest approach before merging. 

This might also buy some time to address feedback around the new test. However, if the test could be a bit brittle or not provide real value, I think we could look to omit it and land this PR without it to make sure it is in 6.7. Just my two cents.

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (kevin940726) [reply test/e2e/specs/editor/blocks/cover.spec.js]: > There's also another issue where certain elements/blocks behind the navigation overlay can still be interacted with.

Yeah, perhaps we need to think more about this as this might become an accessibility concern. That doesn't have to be done in this PR though!

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (getdave) [reply test/e2e/specs/editor/blocks/cover.spec.js]: This doesn't have any accessible semantics. We could use a data-test-id that is only added in certain environments?

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:88]: This does feel unusual. Is there any precedent?

**Comment by renatho** [comment]
Rebased the PR now with [the revert](https://github.com/WordPress/gutenberg/pull/66074) merged.

File: packages/block-library/src/navigation/edit/index.js
Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js]: Makes sense! Removed here: https://github.com/WordPress/gutenberg/pull/66093/commits/10e1007896eb98779c09009ec4e3b427d96c26d0. The hook name is already self-explanatory.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:88]: Is this actually needed? I used the testing instructions against `trunk` and was able to replicate the issue but only on the front of the site and not in the Editor.


https://github.com/user-attachments/assets/eb0ef33e-8641-4565-ae07-754052df9dcc

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (renatho) [reply test/e2e/specs/editor/blocks/cover.spec.js]: > Maybe we could do a [.click( { trial: true } )](https://playwright.dev/docs/api/class-locator#locator-click-option-trial) on the elements in question.

Oh! That's a great suggestion! That was what I was trying initially and I didn't find. Updated here:https://github.com/WordPress/gutenberg/pull/66093/commits/0a244ec93d1f8eadd50b50a4689bda26a5d0240e

The only cons is that it needs to wait until the timeout, but I forced it to be 1 second. I think that's fine. WDYT?

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (renatho) [reply test/e2e/specs/editor/blocks/cover.spec.js]: After https://github.com/WordPress/gutenberg/commit/0a244ec93d1f8eadd50b50a4689bda26a5d0240e, we removed this part, but we still have one `locator` inside the test, similar to the other tests. I hope that's fine to not need change the markup. WDYT?

File: packages/block-library/src/navigation/edit/index.js
Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: Unfortunately, it's really needed because also happens in the editor.

Screenshot 2024-10-15 at 09 51 06

Since it's the behavior in the frontend, I think it would be fine to have the same behavior replicated in the editor.

> I used the testing instructions against trunk and was able to replicate the issue but only on the front of the site and not in the Editor.

I noticed sometimes some kind of cache. I didn't dig into to figure out, but sometimes when it delayed I restarted the build and restarted the browser, also marking the "Disable cache" in my network tab.

Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:88]: Are there specific steps to replicate that state? I tried following the PR instructions but it didn't manifest. Maybe it was prior to the rebase?

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: Ah, sorry! I forgot to add to the instructions to apply the https://github.com/WordPress/gutenberg/pull/66074 before testing.

After the rebase, if you comment the style change, you will see the error happening.

Comment (andrewserong) [reply packages/block-library/src/navigation/edit/index.js:88]: > This does feel unusual. Is there any precedent?

I agree it feels unusual and I couldn't find a precedent with other core blocks. That said, in this case one of the things I kind of like about adding and removing classes via the `useEffect` and attaching to the node, is that the logic is contained within the block itself, and the editor doesn't need to be aware of it — so, we're not handling the state of the modal within the block-editor state for example.

File: packages/block-library/src/navigation/edit/index.js
Comment (andrewserong) [inline packages/block-library/src/navigation/edit/index.js]: Should there be a cleanup function in this `useEffect` to be safe? (Returning a function that removes the class if the component is unmounted). Otherwise if someone deletes the Navigation block while the modal is open, the class persists:

https://github.com/user-attachments/assets/072e3df5-70f8-459f-bdfe-f2646ff28373

File: packages/block-library/src/navigation/edit/index.js
Comment (andrewserong) [inline packages/block-library/src/navigation/edit/index.js]: Tiny nit: This appears to be traversing up from the current ref to get to `html`. For consistency with how the `view.js` approach handles it, would it be worth using `documentElement` instead? I.e. something like:

```
const htmlElement = navRef.current.ownerDocument.documentElement;
```

**Review by andrewserong — COMMENTED** [review]
This is working well for me in testing. I agree it's a little unusual to be adding classes at the root node level from within a block in the editor, but I also like that it creates consistency with what's happening on the frontend. Just added a couple of tiny comments (one nit, and one where I think we need to do a tiny bit of cleanup).

Curious to hear what other folks think, too! It's a tiny bit unfortunate that the classname is so generic (`has-modal-open` vs `wp-block-navigation--has-modal-open` or something like that), but that's not really in scope of this PR.

File: packages/block-library/src/navigation/edit/index.js
Comment (aaronrobertshaw) [reply packages/block-library/src/navigation/edit/index.js]: Good call. +1 for cleanup.

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (aaronrobertshaw) [inline test/e2e/specs/editor/blocks/cover.spec.js:292]: Was this timing simply arbitrary or was there some metric driving it? 

@kevin940726 in your testing wisdom do you have any better ideas?

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: I was literally about to type this and you already spotted. Yes we should ensure we clean up here 👍

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js]: Good point! Thank you for the suggestion folks! Added here: https://github.com/WordPress/gutenberg/pull/66093/commits/0850938415896b525b44fb9c7eaf0b5dcad7d7ca

File: packages/block-library/src/navigation/edit/index.js
Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js]: Sure! Much better! https://github.com/WordPress/gutenberg/pull/66093/commits/efe0a9ea417413c76635ba1e0fda5c52ff551ec9

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (renatho) [reply test/e2e/specs/editor/blocks/cover.spec.js:292]: It wasn't metric-driven. I just used a value that I considered "acceptable". I'm looking forward to seeing if @kevin940726 has ideas to improve this. :)

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: Will `htmlElement` always be defined. For example if the hook bales out early on the `if ( ! navRef.current ) {` check could we conceivably end up with an error.

```suggestion
			htmlElement?.classList.remove( 'has-modal-open' );
```

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js]: Good point! I thought it wouldn't be an issue because the `useEffect` wouldn't have its cleanup if the `navRef.current` is not defined.

But thinking more, since it's about a ref I think it could be changed between the render and the cleanup, not triggering a new `useEffect` based on this change.

Your suggestion was committed here: https://github.com/WordPress/gutenberg/pull/66093/commits/eaabc88c16f362d45c9e839d243779f2f971798e

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (aaronrobertshaw) [reply test/e2e/specs/editor/blocks/cover.spec.js]: I'm not sure. Ultimately, I like having a test to guard against regressions here. So to me if that is the cost of having that protection I could live with it but I don't have strong opinions. I'd like to defer the final call there to others that have more knowledge around Playwright etc.

I'll resolve this thread as the inline comment for the code in question will catch the final decision.

File: test/e2e/specs/editor/blocks/cover.spec.js
Comment (aaronrobertshaw) [reply test/e2e/specs/editor/blocks/cover.spec.js]: Is this something that could be addressed in a code quality follow-up to this PR?

File: packages/block-library/src/navigation/edit/index.js
Comment (aaronrobertshaw) [reply packages/block-library/src/navigation/edit/index.js:88]: One downside is that this sort of makes the Navigation block dependent on the Cover block adjusting its own z-index values. It's not so much that the coupling exists but that it is rather hidden to anyone working on the blocks in the future (hence the importance of the e2e test I guess).

With a lack of better alternatives, I'd be on board with progressing this fix and iterating again if needed.

Comment (andrewserong) [reply packages/block-library/src/navigation/edit/index.js:88]: > With a lack of better alternatives, I'd be on board with progressing this fix and iterating again if needed.

I was just thinking the same. +1 from me for progressing with this and revisiting further down the track.

**Review by aaronrobertshaw — APPROVED** [review]
Appreciate the continued iteration here @renatho 👍 

I've given this another smoke test after the latest round of updates and it all seems to still be testing well for me.

I think there's only two questions we need to answer before merging:

1. Are we okay with the addition of `has-modal-open` class in the editor to match the frontend?
2. Is the arbitrary timeout value ok in the e2e test?

For me, I'd rather see bugs fixed than having arguably imperfect solutions blocked when they don't appear to negatively affect much else. So on those grounds I'll give this an approval.

I don't hold that opinion strongly so it would be prudent to wait for a second approval before merging. On that front, it looks like consensus to move forward is [starting to coalesce](https://github.com/WordPress/gutenberg/pull/66093#discussion_r1804009204).

I am pretty confident though we can address anything that needs it, in a follow-up.

**Review by andrewserong — APPROVED** [review]
Thanks for laying out your thoughts @aaronrobertshaw, that matches my thinking, too!

> I think there's only two questions we need to answer before merging:
> Are we okay with the addition of has-modal-open class in the editor to match the frontend?

In this case, while it is a little unusual, the change is contained in the sense that it only happens when the modal is opened within a navigation block, and the added rules related to that classname are fairly benign. While it'd be good to avoid this approach if possible, right now, this seems like the most expedient fix, and I haven't encountered any conflicts or issues with the approach.

> Is the arbitrary timeout value ok in the e2e test?

For now, I'd think so. It's good that we have some e2e test coverage, and tests can be followed up on in subsequent PRs.

So, all up, on balance I think the UX benefits of this fix (that cover blocks aren't popping in randomly from the perspective of a user) outweigh the code quality concerns we've encountered of a block possibly overreaching by updating classnames on the document wrapper.

A +1 to landing from me! Nice work @renatho, thanks for all the back and forth on this.

**Comment by aaronrobertshaw** [comment]
Thanks for weighing in again @andrewserong 👍 

I'll take the liberty to get this merged for @renatho then. 

> Nice work @renatho, thanks for all the back and forth on this.

+1

File: packages/block-library/src/navigation/edit/index.js
Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: For the future, I think the ideal solutions would be:

> 1. Having the menu opened in a totally different context would be the ideal solution to avoid any stacking context issues. But besides it being a big change, it could impact other things, like not inheriting styles from the current context and creating other unexpected style issues.
> 2. Change the order of the cover elements (just invert the image and the color effect span). With this, we could remove the z-index of the elements, and it would solve the issue. The problem would be the block deprecation and a complex code to maintain both versions to not break the styles on the frontend until the user re-saves the cover block.

The problem with those solutions is mostly backward compatibility.

Comment (aaronrobertshaw) [reply packages/block-library/src/navigation/edit/index.js:88]: > The problem would be the block deprecation

I'm missing what the problem with a block deprecation would be. Can you expand on this aspect?

> a complex code to maintain both versions to not break the styles

We unfortunately have to keep a fair bit of CSS around for deprecated versions of blocks. I'm guessing what would be required here isn't a lot.

As for the block's render callback, it doesn't look overly complex at a glance. So it could likely absorb any slight change in logic for BC.

**Comment by getdave** [comment]
Thanks everyone 👍

File: packages/block-library/src/navigation/edit/index.js
Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: Hi @aaronrobertshaw! I could be missing something, but I think for deprecations changing the markup it looks "broken" (with the warning) in the editor and the user needs to click in a button to fix the block, right?

So I thought that it would be better to avoid it if we could for the cover block because is very common and it is a wrapper component.

For the render callback, I'm not sure if I understand what you mean. Do you mean to invert the position of the elements dynamically only? If so, I'm not sure if it's a good idea because we need the same in the `edit`. Unless we change the order in the `edit` for the editor, and change in the render callback for the frontend, although I think it would be a weird code for maintenance.

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: I just checked I was wrong about the deprecation and it's transparent for the user. I'll raise another PR to improve it more. And the solution of the `has-modal-open` will be only for backward compatibility. =)

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: I raised [this PR](https://github.com/WordPress/gutenberg/pull/66249). I'd appreciate if someone could review it to improve the solution.

I also tried to update the fixtures through `npm run fixtures:regenerate`, but I received some errors. If someone could explain me what is happening and how I could fix, it would also be very helpful to save some time. =)

Comment (aaronrobertshaw) [reply packages/block-library/src/navigation/edit/index.js:88]: > I just checked I was wrong about the deprecation and it's transparent for the user

Yep, the only real sign within the editor should be a console message stating the block was migrated.

> And the solution of the has-modal-open will be only for backward compatibility

This PR says it should be in the 19.6 release. So if we do go with the alternative solution soon I don't think there'll be any need to keep the `.has-modal-open` class styles around for BC. The change introducing it won't have been released in other words.

I'll comment to that effect when I get to review the alternate PR.

> I also tried to update the fixtures through npm run fixtures:regenerate, but I received some errors.

Without looking I can't say if this is the reason or not but another misconception about deprecations is that they are chained together to migrate one deprecated version to the next, all the way to the latest version.

Instead, all deprecations need to be updated (if required) to migrate that deprecated version to the latest. We can continue the discussion over on https://github.com/WordPress/gutenberg/pull/66249 as needed.

Thanks for spinning it up 🚀

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: > This PR says it should be in the 19.6 release. So if we do go with the alternative solution soon I don't think there'll be any need to keep the .has-modal-open class styles around for BC. The change introducing it won't have been released in other words.

It's good to have that anyway, so it's fixed everywhere not needing a manual save in all the broken posts/templates. :)

Comment (aaronrobertshaw) [reply packages/block-library/src/navigation/edit/index.js:88]: Wasn't the original issue only in the editor? If that's the case, the deprecation will migrate and fix that display without needing further styles.

Comment (renatho) [reply packages/block-library/src/navigation/edit/index.js:88]: Connecting the points, I sent an answer here: https://github.com/WordPress/gutenberg/pull/66249#discussion_r1810669152

**Pull request merged**
